clippy_lints/src/manual_unwrap_or_default.rs

```diff
@@ -131,6 +131,17 @@ fn handle<'tcx>(cx: &LateContext<'tcx>, if_let_or_match: IfLetOrMatch<'tcx>, exp
         },
     };
 
+    // We check if the expression is not a None variant
+    if let Some(none_def_id) = cx.tcx.lang_items().option_none_variant() {
```

Comment:
I think the issue is fairly general and doesn't only happen with a hardcoded `None` specifically, e.g.:
```rs
let _ = if let Some(x) = "1".parse().ok() {
  x
} else {
  i32::default()
};
```
This also has a broken suggestion, for the same reason as `None` -- there's an inference variable created in the condition that is only constrained to be equal to a specific type in the else case, so removing the else block breaks inference, and that inference variable could be created in many different ways.

It might be possible to use [`expr_type_is_certain`](https://doc.rust-lang.org/stable/nightly-rustc/clippy_utils/ty/fn.expr_type_is_certain.html) on the `condition`, though from my experience, it has a lot of false negatives, so maybe this regresses too many true positive cases. Would still be worth a try though.

Reply:
I added the support for all the expression that don't have a specified type.